st2api/tests/unit/controllers/v1/test_alias_execution.py

```diff
@@ -175,17 +190,76 @@ def test_match_and_execute(self, mock_request):
+        }
+
         # Command matches - should result in action execution
         data = copy.deepcopy(base_data)
         data['command'] = 'run date on localhost'
         resp = self.app.post_json('/v1/aliasexecution/match_and_execute', data)
         self.assertEqual(resp.status_int, 201)
-        self.assertEqual(resp.json['execution']['id'], str(EXECUTION['id']))
-        self.assertEqual(resp.json['execution']['status'], EXECUTION['status'])
+        self.assertEqual(len(resp.json['results']), 1)
+        self.assertEqual(resp.json['results'][0]['execution']['id'], str(EXECUTION['id']))
+        self.assertEqual(resp.json['results'][0]['execution']['status'], EXECUTION['status'])
 
         expected_parameters = {'cmd': 'date', 'hosts': 'localhost'}
         self.assertEquals(mock_request.call_args[0][0].parameters, expected_parameters)
 
+    @mock.patch.object(action_service, 'request',
+                       return_value=(None, EXECUTION))
+    def test_match_and_execute_matches_one_multiple_match(self, mock_request):
+        base_data = {
+            'source_channel': 'chat',
+            'notification_route': 'hubot',
+            'user': 'chat-user'
+        }
+
+        # Command matches multiple times - should result in multiple action execution
+        data = copy.deepcopy(base_data)
+        data['command'] = 'JKROWLING-4 is a duplicate of JRRTOLKIEN-24 which is a duplicate of DRSEUSS-12'
+        resp = self.app.post_json('/v1/aliasexecution/match_and_execute', data)
+        self.assertEqual(resp.status_int, 201)
+        self.assertEqual(len(resp.json['results']), 2)
+        self.assertEqual(resp.json['results'][0]['execution']['id'], str(EXECUTION['id']))
+        self.assertEqual(resp.json['results'][0]['execution']['status'], EXECUTION['status'])
+        self.assertEqual(resp.json['results'][1]['execution']['id'], str(EXECUTION['id']))
+        self.assertEqual(resp.json['results'][1]['execution']['status'], EXECUTION['status'])
+
+        # The mock object only stores the parameters of the _last_ time it was called, so that's
+        # what we assert on. Luckily re.finditer() processes groups in order, so if this was the
+        # parameters to the mock object, we have _also_ called it with:
+        #
+        # {'issue_key': 'JRRTOLKIEN-24'}
+        #
+        # We've also already checked the results array
+        #
+        expected_parameters = {'issue_key': 'DRSEUSS-12'}
+        self.assertEquals(mock_request.call_args[0][0].parameters, expected_parameters)
+
+    @mock.patch.object(action_service, 'request',
+                       return_value=(None, EXECUTION))
+    def test_match_and_execute_matches_many_multiple_match(self, mock_request):
+        base_data = {
+            'source_channel': 'chat',
+            'notification_route': 'hubot',
+            'user': 'chat-user'
+        }
+
+        # Command matches multiple times - should result in multiple action execution
```

Comment:
I'm a bit confused about this comment and test. Why do we expect and error here? Is `match_multiple` set to false or something else is going on?

Edit: I see, it looks like the reason is there are two overlapping aliases, right?

Reply:
No, in the `test_match_and_execute_matches_many_multiple_match` test there are two aliases that matched. The intent of my code, and of the original code, was to only ever match *one* alias, never more than one.

However, that one matched alias could match the command multiple times, and cause multiple actions to be executed.

Pseudocode looks like:

```python
matched_aliases = []
for alias in all_aliases:
    for format in alias.formats:
        for representation in format.representations:
            if representation.match(command):
                matches.append(alias)

assert len(matched_aliases) == 1

alias = matched_aliases.first()

results = []
for match_parameters in alias.match(command):
    results.append(alias.action.execute(match_parameters))

return results
```

----------------

Now that we are returning an array of match results, we could also allow multiple aliases to match. That pseudocode would look like:

```python
matched_aliases = []
for alias in all_aliases:
    for format in alias.formats:
        for representation in format.representations:
            if representation.match(command):
                matched_aliases.append(alias)

results = []
for matched_alias in matched_aliases:
    for match_parameters in matched_alias.match(command):
        results.append(matched_alias.action.execute(match_parameters))

return results
```

Is matching multiple aliases acceptable?

Reply:
Just saw your edit: yep, that's why. Should I match multiple aliases since I'm changing the semantics of `match_and_execute` anyway?

Reply:
I think it's fine the way it is right now :)